src/java.base/share/classes/java/lang/foreign/Linker.java

```diff
@@ -60,64 +61,340 @@
+ * defined in C libraries (native functions). Suppose we wish to downcall from Java to the {@code strlen} function
+ * defined in the standard C library:
+ * {@snippet lang = c:
+ * size_t strlen(const char *s);
+ * }
+ * A downcall method handle that exposes {@code strlen} is obtained, using the native linker, as follows:
+ *
+ * {@snippet lang = java:
+ * Linker linker = Linker.nativeLinker();
+ * MethodHandle strlen = linker.downcallHandle(
+ *     linker.defaultLookup().find("strlen").get(),
+ *     FunctionDescriptor.of(JAVA_LONG, ADDRESS)
+ * );
+ * }
+ *
+ * Note how the native linker also provides access, via its {@linkplain #defaultLookup() default lookup},
+ * to the native functions defined by the C libraries loaded with the Java runtime. Above, the default lookup
+ * is used to search the address of the {@code strlen} native function. That address is then passed, along with
+ * a <em>platform-dependent description</em> of the signature of the function expressed as a
+ * {@link FunctionDescriptor} (more on that below) to the native linker's
+ * {@link #downcallHandle(MemorySegment, FunctionDescriptor, Option...)} method.
+ * The obtained downcall method handle is then invoked as follows:
+ *
+ * {@snippet lang = java:
+ * try (Arena arena = Arena.openConfined()) {
+ *     MemorySegment str = arena.allocateUtf8String("Hello");
+ *     long len          = strlen.invoke(str);  // 5
+ * }
+ * }
+ * <h3 id="describing-c-sigs">Describing C signatures</h3>
+ *
+ * When interacting with the native linker, clients must provide a platform-dependent description of the signature
+ * of the C function they wish to link against. This description, a {@link FunctionDescriptor function descriptor},
+ * defines the layouts associated with the parameter types and return type (if any) of the C function.
+ * <p>
+ * Scalar C types such as {@code bool}, {@code int} are modelled as {@linkplain ValueLayout value layouts}
+ * of a suitable carrier. The mapping between a scalar type and its corresponding layout is dependent on the ABI
+ * implemented by the native linker. For instance, the C type {@code long} maps to the layout constant
+ * {@link ValueLayout#JAVA_LONG} on Linux/x64, but maps to the layout constant {@link ValueLayout#JAVA_INT} on
+ * Windows/x64. Similarly, the C type {@code size_t} maps to the layout constant {@link ValueLayout#JAVA_LONG}
+ * on 64-bit platforms, but maps to the layout constant {@link ValueLayout#JAVA_INT} on 32-bit platforms.
+ * <p>
+ * Composite types are modelled as {@linkplain GroupLayout group layouts}. More specifically, a C {@code struct} type
+ * maps to a {@linkplain StructLayout struct layout}, whereas a C {@code union} type maps to a {@link UnionLayout union
+ * layout}. Depending on the ABI implemented by the native linker, additional {@linkplain MemoryLayout#paddingLayout(long) padding}
+ * member layouts might be required to conform to the size and alignment constraint of a composite type definition in C.
```

Comment:
I think the comment on padding here is maybe a bit too vague. I think it would be better to say explicitly that padding is not added to group layouts automatically, and has to be manually specified.

(On that note, maybe we want to add a factory in MemoryLayout that basically exposes `Utils.computePaddedStruct`)